### ISSUE TITLE
Phrasing of E0507 is ambiguous with regards to moved variables

### ISSUE THREAD

**Issue by dvogel** [rust-lang/rust#101459]
Given the following code: 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cf3bd67f07db7324126bcade4e68471f

```rust
fn printlen(s: String) {
    println!("s.len() = {}", s.len());
}

fn goodfunc<F: FnOnce()>(callback: F) {
    callback();
}

fn badfunc<F: Fn()>(callback: F) {
    callback();
}

fn main() {
    let a = String::from("xyz");
    let b = a.clone();
    
    goodfunc(move || {
        printlen(a);
    });
    
    badfunc(move || {
        printlen(b);
    });
}
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[[E0507]](https://doc.rust-lang.org/stable/error-index.html#E0507): cannot move out of `b`, a captured variable in an `Fn` closure
  --> src/main.rs:23:18
   |
16 |       let b = a.clone();
   |           - captured outer variable
...
22 |       badfunc(move || {
   |  _____________-
23 | |         printlen(b);
   | |                  ^ move occurs because `b` has type `String`, which does not implement the `Copy` trait
24 | |     });
   | |_____- captured by this `Fn` closure

For more information about this error, try `rustc --explain E0507`.
error: could not compile `playground` due to previous error
```

This phrasing has favored brevity over clarity, overloading the captured variable name:

> cannot move out of `b`, a captured variable in an `Fn` closure

The comma is easy to miss but it is doing a lot of the work here. I believe there are multiple valid English-language interpretations of that comma. The one I now understand to be correct:

> The variable `b` cannot be moved out of the `Fn` closure, where it has already been captured.

The way I originally interpreted it:

> The variable `b` is captured in the `Fn` closure but that would require an invalid move out of `main`.

To properly interpret the existing error the reader has to intuit that the compiler considers there to be an implicit variable of the same name `b` existing inside the closure. That may be technically true but the explicit reference to the variable name draws the reader's attention to the only explicit declaration of `b`, in `main`. I believe either of these phrasings avoid the ambiguity while also focusing the user's attention on the problematic move:

> The capture of `b` cannot be moved out of the `Fn`

> The variable `b` cannot be moved out of `Fn` because the closure may be called multiple times.

**Comment by steffahn** [comment]
> To properly interpret the existing error the reader has to intuit that the compiler considers there to be an implicit variable of the same name `b` existing inside the closure.

Yes, that’s the crucial information to understand the error message correctly in my opinion, too.

---

> I believe either of these phrasings avoid the ambiguity while also focusing the user's attention on the problematic move:

I don’t like either of the “improvements” you suggested, in particular it must still say “`Fn` closure” in my opinion.

---

Regarding analysis of the current phrasing, I would interpret

> cannot move out of `b`, a captured variable in an `Fn` closure

as
> cannot move out of `b`, which is a captured variable in an `Fn` closure

i.e.
> cannot move out of `b`. `b` is a captured variable in an `Fn` closure

and go with the “[there is] an implicit variable of the same name `b` existing inside the closure” interpretation as you explained above.

However, on second thought, I *think* what the error message is ***actually*** trying to say is

> [you] cannot more out of `b` because you cannot move out a captured variable in an `Fn` closure.

or

> cannot move out a captured variable in an `Fn` closure, and by the way, the variable this error applies to is called `b`

---

In light of this last interpretation, perhaps we should simply remove the mention of the variable name `b` in the main error message. E.g. something like
```

error[E0507]: cannot move out of a captured variable in an `Fn` closure
  --> src/main.rs:23:18
   |
16 |     let b = a.clone();
   |         - this variable is being captured
...
22 |     badfunc(move || {
   |             ------- the variable `b` is captured by this `Fn` closure
23 |         printlen(b);
   |                  ^ move occurs because `b` has type `String`, which does not implement the `Copy` trait
   |
note: you can only move out of captured variables in `FnOnce` closures, but never in `FnMut` closures or `Fn` closures

For more information about this error, try `rustc --explain E0507`.
```
I’ve added a `note` to make clear that there’s no specific problem with this particular variable being captured by this particular `Fn` closure, but that it’s impossible ***in general*** to move out of a captured variable in a `Fn` closure.

What do you think about the above error message?

Maybe also add an additional note for *why* the closure is an `Fn` closure (in this case pointing to the type signature of `badfunc` that made it this way.

If the question *why* moving out of captured variables in `Fn` closures is impossible regularly comes up, then there should be an addition about this case to the `rustc --explain E0507` output, or perhaps a new error code specifically for this kind of error, as opposed to the general “E0507” whose [description](https://doc.rust-lang.org/nightly/error_codes/E0507.html) starts with the not very fitting (in this case) “A borrowed value was moved out.”

**Comment by dvogel** [comment]
I think the error message as you've rephrased it is an improvement over the one I ran into. I definitely like the `FnOnce` mention being a `note` rather than a hint directly in the error because sometimes the user won't also be the author who chose the closure type. That was the case for me when I ran into this. I was using [this method from gtk-rs](https://gtk-rs.org/gtk3-rs/stable/latest/docs/gtk/prelude/trait.WidgetExt.html#tymethod.connect_realize) that conceptually should only be called once but is declared as an `Fn` (I'm guessing due to a weird GTK quirk where it can occur multiple times). So the `FnOnce` mention is definitely important but making it a note feels like a nice reminder that there's complications with closure types rather than an indication of an errant choice made by the user.

While I think the note about `Fn` vs `FnOnce` closures is helpful in it's current form, I think it would be more obviously related to the error if the error more directly highlighted the dual nature of that line. It is both where the variable is moved out of the closure _and_ where it is re-used. 

In [a more traditional case of re-use after move](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fbb21a096b105f80a137564969e5dd93):

```
fn takes_ownership(s: String) {
    println!("My string: {s}");
}

fn main() {
    let a = String::from("xyz");
    takes_ownership(a);
    takes_ownership(a);
}
```

The user is told both where the variable was moved and where it is re-used:

```
Compiling playground v0.0.1 (/playground)
error[[E0382]](https://doc.rust-lang.org/stable/error-index.html#E0382): use of moved value: `a`
 --> src/main.rs:8:21
  |
6 |     let a = String::from("xyz");
  |         - move occurs because `a` has type `String`, which does not implement the `Copy` trait
7 |     takes_ownership(a);
  |                     - value moved here
8 |     takes_ownership(a);
  |                     ^ value used here after move

For more information about this error, try `rustc --explain E0382`.
error: could not compile `playground` due to previous error
```

While these are distinct errors due to the nature of the value that is moved, I think they share the same two part structure. Two separate things have been done with the variable. It was moved out and it was also re-used (in a hypothetical subsequent call to the `Fn` closure). So the error for line 23 could be something more like:

> value moved here and used here in a subsequent `Fn` closure call.

With this phrasing the link is much more easily made between the error calling this a `Fn` closure and the note contrasting with a `FnOnce` closure.

The phrase about `String` not implementing `Copy` seem more appropriate for the annotation on line 22, mirroring the structure of E0382 where that same phrase annotates the declaration on line 6.

**Comment by Dhghomon** [comment]
I'm putting together an example on closures and came across this issue after noticing the error message. The example involves recreating `Option` as a simpler concrete type in order to teach how `unwrap_or_else` works and why it takes an FnOnce instead of something else. Then down in main it captures a String which would make it into an FnOnce, and thus doesn't compile.

```
#[derive(Clone)]
pub enum StringOption {
    Some(String),
    None,
}

impl StringOption {
    pub fn unwrap_or_else<F>(self, f: F) -> String
    where
        F: Fn() -> String,
    {
        match self {
            StringOption::Some(x) => x,
            StringOption::None => f(),
        }
    }
}

fn main() {
    let default = "false".to_string();
    let something = StringOption::Some("Some data".to_string());

    println!("{}", something.unwrap_or_else(|| { default }));
}
```

The closure inside `.unwrap_or_else()` here isn't actually an `Fn` closure, it's just a method that is supposed to take one. So here it would be nice to see a message along the lines of "function StringOption::unwrap_or_else() takes a closure that implements Fn, found a closure that only implements FnOnce".
